TYPE=['WT','KO']
SAMPLES = ['1', '2', '3']
READS=['1','2']

rule all:
    input:
        expand("{type}_{sample}_R{read_no}.fastq.gz", type=TYPE,sample=SAMPLES,read_no=READS),
        expand("{type}_{sample}.startlog",type=TYPE,sample=SAMPLES),
        expand("{type}_{sample}Aligned.sortedByCoord.out.bam",type=TYPE,sample=SAMPLES),
        expand("{type}_{sample}Aligned.sortedByCoord.out.pair.bam",type=TYPE,sample=SAMPLES),
        expand("{type}_{sample}Aligned.sortedByCoord.out.pair.fix.bam",type=TYPE,sample=SAMPLES),
        expand("{type}_{sample}Aligned.sortedByCoord.out.pair.fix.sort.bam",type=TYPE,sample=SAMPLES),
        expand("{type}_{sample}Aligned.sortedByCoord.out.pair.fix.sort.dedup.bam",type=TYPE,sample=SAMPLES),
        expand("{type}_{sample}.tsv", type=TYPE,sample=SAMPLES)

rule fastp:
    input:
        reads1="{type}_{sample}_R1_001.fastq.gz",
        reads2="{type}_{sample}_R2_001.fastq.gz"
    output:
        reads1out="{type}_{sample}_R1.fastq.gz",
        reads2out="{type}_{sample}_R2.fastq.gz",
        html="{type}_{sample}.html",
        json="{type}_{sample}.json"
    threads:
        4
    shell:
        "./fastp -i {input.reads1} -I {input.reads2} -o {output.reads1out} -O {output.reads2out} -D3 -h {output.html} -j {output.json} -f 15 -F 15 -t 3 -T 3"

rule star:
    input:
        reads1="{type}_{sample}_R1.fastq.gz",
        reads2="{type}_{sample}_R2.fastq.gz"
    output:
        logfile="{type}_{sample}.startlog",
        outfile1="{type}_{sample}Aligned.sortedByCoord.out.bam",
        #outfile2="{type}_{sample}Log.final.out",
        #outfile3="{type}_{sample}Log.out",
        #outfile4="{type}_{sample}Log.progress.out",
        #outfile5="{type}_{sample}SJ.out.tab",
    params:
        stem1="{type}_{sample}_star",
        stem2="{type}_{sample}"
    threads:
        4 
    shell:
        "STAR \
        --runThreadN 4 \
        --genomeDir ~/projects/zebrafish/genome/STAR/ \
        --readFilesIn {input.reads1} {input.reads2}  \
        --readFilesCommand zcat \
        --outSAMtype BAM SortedByCoordinate \
        --outFileNamePrefix {params.stem2}  >& {output.logfile}"

rule pair:
    input:
        bam="{type}_{sample}Aligned.sortedByCoord.out.bam"
    output:
        bam="{type}_{sample}Aligned.sortedByCoord.out.pair.bam"
    threads:
        4
    shell:
        "/usr/bin/samtools sort -n -o {output.bam} {input.bam}"

rule fixmate:
    input:
        bam="{type}_{sample}Aligned.sortedByCoord.out.pair.bam"
    output:
        bam="{type}_{sample}Aligned.sortedByCoord.out.pair.fix.bam"
    threads:
        4
    shell:
        "/usr/bin/samtools fixmate -m {input.bam} {output.bam}"

rule resort:
    input:
        bam="{type}_{sample}Aligned.sortedByCoord.out.pair.fix.bam"
    output:
        bam="{type}_{sample}Aligned.sortedByCoord.out.pair.fix.sort.bam"
    threads:
        4
    shell:
        "/usr/bin/samtools sort -o {output.bam} {input.bam}"

rule dedup:
    input:
        bam="{type}_{sample}Aligned.sortedByCoord.out.pair.fix.sort.bam"
    output:
        bam="{type}_{sample}Aligned.sortedByCoord.out.pair.fix.sort.dedup.bam"
    threads:
        4
    shell:
        "/usr/bin/samtools markdup -r {input.bam} {output.bam}"

rule htseq:
    input:
        bam="{type}_{sample}Aligned.sortedByCoord.out.pair.fix.sort.dedup.bam"
    output:
        tsv="{type}_{sample}.tsv"
    threads:
        4
    shell:
        "/usr/bin/samtools view -f 3 -h {input.bam} | htseq-count -f sam - ~/projects/zebrafish/genome/Danio_rerio.GRCz11.105.gtf -a 10 --order=pos -s reverse > {output.tsv}" 




